File: src/transcripts/parser.ts
/**
 * Transcript Parser
 * Parse JSONL transcript files from Claude Code sessions
 */

import type { ContentBlock, RawTranscriptEntry, TranscriptLine } from './types';

/**
 * Parse a transcript JSONL content string into structured lines
 * @param content - Raw JSONL content
 * @returns Array of parsed transcript lines
 */
export function parseTranscript(content: string): TranscriptLine[] {
  const lines = content.split('\n');
  const results: TranscriptLine[] = [];

  for (let i = 0; i < lines.length; i++) {
    const rawLine = lines[i];
    if (!rawLine) continue;
    const line = rawLine.trim();
    if (!line) continue;

    try {
      const parsed = JSON.parse(line) as RawTranscriptEntry;
      const transcriptLine = normalizeTranscriptLine(parsed, i + 1, line);
      if (transcriptLine) {
        results.push(transcriptLine);
      }
    } catch {
      // Skip malformed JSON lines gracefully
      console.warn(`Skipping malformed JSON at line ${i + 1}`);
    }
  }

  return results;
}

/**
 * Parse a transcript file from disk
 * @param filePath - Path to the JSONL file
 * @returns Array of parsed transcript lines
 */
export async function parseTranscriptFile(filePath: string): Promise<TranscriptLine[]> {
  const file = Bun.file(filePath);
  const exists = await file.exists();

  if (!exists) {
    throw new Error(`Transcript file not found: ${filePath}`);
  }

  const content = await file.text();
  return parseTranscript(content);
}

/**
 * Get a specific line from a transcript file
 * @param filePath - Path to the JSONL file
 * @param lineNumber - 1-based line number
 * @returns The transcript line or null if not found
 */
export async function getTranscriptLine(
  filePath: string,
  lineNumber: number
): Promise<TranscriptLine | null> {
  const file = Bun.file(filePath);
  const exists = await file.exists();

  if (!exists) {
    return null;
  }

  const content = await file.text();
  const lines = content.split('\n');

  if (lineNumber < 1 || lineNumber > lines.length) {
    return null;
  }

  const rawLine = lines[lineNumber - 1];
  if (!rawLine) return null;
  const line = rawLine.trim();
  if (!line) return null;

  try {
    const parsed = JSON.parse(line) as RawTranscriptEntry;
    return normalizeTranscriptLine(parsed, lineNumber, line);
  } catch {
    return null;
  }
}

/**
 * Extract text content from a transcript line
 * @param line - The transcript line
 * @returns Extracted text or null if no text content
 */
export function extractTextContent(line: TranscriptLine): string | null {
  if (!line.message) return null;

  const { content } = line.message;

  if (typeof content === 'string') {
    return content;
  }

  if (Array.isArray(content)) {
    const textParts: string[] = [];

    for (const block of content) {
      if (block.type === 'text' && block.text) {
        textParts.push(block.text);
      } else if (block.type === 'tool_use' && block.name) {
        textParts.push(`[Tool: ${block.name}]`);
      } else if (block.type === 'tool_result' && block.text) {
        textParts.push(block.text);
      }
    }

    return textParts.length > 0 ? textParts.join('\n') : null;
  }

  return null;
}

/**
 * Normalize a raw transcript entry into a structured TranscriptLine
 */
function normalizeTranscriptLine(
  raw: RawTranscriptEntry,
  lineNumber: number,
  rawString: string
): TranscriptLine | null {
  // Determine the type
  let type: TranscriptLine['type'] = 'system';

  if (raw.type) {
    const rawType = raw.type.toLowerCase();
    if (
      rawType === 'user' ||
      rawType === 'assistant' ||
      rawType === 'file-history-snapshot' ||
      rawType === 'system' ||
      rawType === 'summary' ||
      rawType === 'progress'
    ) {
      type = rawType as TranscriptLine['type'];
    }
  } else if (raw.message?.role) {
    type = raw.message.role as TranscriptLine['type'];
  }

  // Ensure required fields have defaults
  const uuid = raw.uuid || `line-${lineNumber}`;
  const sessionId = raw.sessionId || 'unknown';
  const timestamp = raw.timestamp || new Date().toISOString();
  const cwd = raw.cwd || '';

  return {
    lineNumber,
    type,
    uuid,
    parentUuid: raw.parentUuid ?? null,
    sessionId,
    timestamp,
    cwd,
    version: raw.version,
    gitBranch: raw.gitBranch,
    slug: raw.slug,
    message: raw.message,
    toolUseResult: raw.toolUseResult,
    raw: rawString,
    // Additional fields for system/progress messages
    subtype: raw.subtype,
    data: raw.data as Record<string, unknown> | undefined,
    summary: raw.summary,
    hookInfos: raw.hookInfos as Array<{ command: string }> | undefined,
    hookErrors: raw.hookErrors,
    hookCount: raw.hookCount,
  };
}

/**
 * Extract all tool uses from a transcript
 * @param lines - Parsed transcript lines
 * @returns Array of tool use information
 */
export function extractToolUses(lines: TranscriptLine[]): Array<{
  lineNumber: number;
  toolName: string;
  input: Record<string, unknown>;
}> {
  const toolUses: Array<{
    lineNumber: number;
    toolName: string;
    input: Record<string, unknown>;
  }> = [];

  for (const line of lines) {
    if (!line.message?.content || typeof line.message.content === 'string') {
      continue;
    }

    for (const block of line.message.content as ContentBlock[]) {
      if (block.type === 'tool_use' && block.name) {
        toolUses.push({
          lineNumber: line.lineNumber,
          toolName: block.name,
          input: block.input || {},
        });
      }
    }
  }

  return toolUses;
}

/**
 * Get conversation thread from a specific line (following parentUuid chain)
 * @param lines - All transcript lines
 * @param targetUuid - UUID of the line to get thread for
 * @returns Array of lines in the conversation thread
 */
export function getConversationThread(
  lines: TranscriptLine[],
  targetUuid: string
): TranscriptLine[] {
  const uuidToLine = new Map<string, TranscriptLine>();

  for (const line of lines) {
    uuidToLine.set(line.uuid, line);
  }

  const thread: TranscriptLine[] = [];
  let current = uuidToLine.get(targetUuid);

  while (current) {
    thread.unshift(current);
    if (current.parentUuid) {
      current = uuidToLine.get(current.parentUuid);
    } else {
      break;
    }
  }

  return thread;
}

File: transcript-tui-rs/crates/transcript-core/src/parser.rs
//! JSONL parsing utilities for transcript files

use crate::types::{LineType, ParsedLine, TranscriptLine};
use std::io::{BufRead, BufReader};
use std::path::Path;

/// Parse a single JSONL line into a TranscriptLine
pub fn parse_line(raw: &str, line_number: i64, id: i64) -> Option<TranscriptLine> {
    let parsed: ParsedLine = serde_json::from_str(raw).ok()?;

    let line_type = match parsed.r#type.as_deref() {
        Some("user") => LineType::User,
        Some("assistant") => LineType::Assistant,
        Some("system") => LineType::System,
        Some("summary") => LineType::Summary,
        Some("progress") => LineType::Progress,
        Some("file-history-snapshot") => LineType::FileHistorySnapshot,
        _ => LineType::Unknown,
    };

    // Extract content from message
    let content = if let Some(msg) = &parsed.message {
        Some(msg.content.as_text())
    } else if let Some(summary) = &parsed.summary {
        Some(summary.clone())
    } else {
        None
    };

    let role = parsed.message.as_ref().map(|m| m.role.clone());
    let model = parsed.message.as_ref().and_then(|m| m.model.clone());

    Some(TranscriptLine {
        id,
        line_number,
        line_type,
        uuid: parsed.uuid.unwrap_or_default(),
        parent_uuid: parsed.parent_uuid,
        session_id: parsed.session_id.unwrap_or_default(),
        timestamp: parsed.timestamp.unwrap_or_default(),
        cwd: parsed.cwd,
        slug: parsed.slug,
        role,
        model,
        content,
        raw: raw.to_string(),
        turn_id: None,
        turn_sequence: None,
        session_name: None,
    })
}

/// Parse a JSONL file into transcript lines
pub fn parse_file<P: AsRef<Path>>(path: P) -> std::io::Result<Vec<TranscriptLine>> {
    let file = std::fs::File::open(path)?;
    let reader = BufReader::new(file);
    let mut lines = Vec::new();
    let mut line_number = 1i64;
    let mut id = 1i64;

    for line_result in reader.lines() {
        let raw = line_result?;
        if !raw.trim().is_empty() {
            if let Some(line) = parse_line(&raw, line_number, id) {
                lines.push(line);
                id += 1;
            }
        }
        line_number += 1;
    }

    Ok(lines)
}

/// Extract searchable text from a transcript line
pub fn extract_text(line: &TranscriptLine) -> String {
    let mut parts = Vec::new();

    if let Some(content) = &line.content {
        parts.push(content.clone());
    }

    // Parse raw JSON for additional content
    if let Ok(parsed) = line.parse() {
        // Tool use inputs
        if let Some(msg) = &parsed.message {
            for (_, name, input) in msg.content.tool_uses() {
                parts.push(format!("[Tool: {}]", name));
                if let Some(obj) = input.as_object() {
                    for (key, value) in obj {
                        if let Some(s) = value.as_str() {
                            if s.len() < 500 {
                                parts.push(format!("{}: {}", key, s));
                            }
                        }
                    }
                }
            }
        }

        // Summary text
        if let Some(summary) = &parsed.summary {
            parts.push(summary.clone());
        }
    }

    parts.join("\n")
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_parse_user_line() {
        let raw = r#"{"type":"user","uuid":"abc","sessionId":"123","timestamp":"2024-01-01T00:00:00Z","message":{"role":"user","content":"Hello"}}"#;
        let line = parse_line(raw, 1, 1).unwrap();
        assert_eq!(line.line_type, LineType::User);
        assert_eq!(line.content, Some("Hello".to_string()));
    }

    #[test]
    fn test_parse_assistant_line() {
        let raw = r#"{"type":"assistant","uuid":"def","sessionId":"123","timestamp":"2024-01-01T00:00:01Z","message":{"role":"assistant","content":[{"type":"text","text":"Hi there!"}]}}"#;
        let line = parse_line(raw, 2, 2).unwrap();
        assert_eq!(line.line_type, LineType::Assistant);
        assert_eq!(line.content, Some("Hi there!".to_string()));
    }
}
